k9mail/src/main/java/com/fsck/k9/controller/QresyncExtensionHandler.java

```diff
@@ -0,0 +1,138 @@
+            qresyncParamResponse, FlagSyncHelper flagSyncHelper, SyncHelper syncHelper) throws MessagingException {
+        String folderName = imapFolder.getName();
+        final AtomicInteger headerProgress = new AtomicInteger(0);
+
+        List<ImapMessage> modifiedMessages = qresyncParamResponse.getModifiedMessages();
+        Timber.v("SYNC: Received %d FETCH responses in QRESYNC response", modifiedMessages.size());
+
+        Long cachedSmallestUid = localFolder.getSmallestMessageUid();
+        long smallestLocalUid = cachedSmallestUid == null ? 1 : cachedSmallestUid;
+
+        List<Message> newMessages = new ArrayList<>();
+        List<Message> syncFlagMessages = new ArrayList<>();
+
+        for (Message message : modifiedMessages) {
+            syncHelper.evaluateMessageForDownload(message, folderName, localFolder, imapFolder, account, newMessages,
+                    syncFlagMessages, controller);
+        }
+
+        for (Message message : syncFlagMessages) {
+            long remoteMessageUid = Long.parseLong(message.getUid());
+            if (remoteMessageUid < smallestLocalUid) {
+                continue;
+            }
+
+            flagSyncHelper.processDownloadedFlags(account, localFolder, message);
+
+            headerProgress.incrementAndGet();
+            for (MessagingListener l : controller.getListeners()) {
+                l.synchronizeMailboxProgress(account, folderName, headerProgress.get(), modifiedMessages.size());
+            }
+        }
+
+        for (Message newMessage : newMessages) {
+            remoteMessagesToDownload.add((ImapMessage) newMessage);
+        }
+        Timber.v("SYNC: Received %d new message UIDs in QRESYNC response", remoteMessagesToDownload.size());
+    }
+
+    private void findOldRemoteMessagesToDownload(List<ImapMessage> remoteMessagesToDownload, SyncHelper syncHelper)
+            throws MessagingException {
+        /*At ths point, UIDs and flags for new messages and existing changed messages have been fetched, so all N
+          messages in the local store should correspond to the most recent N messages in the remote store. We still need
+          to download historical messages (messages older than the oldest message in the inbox) if necessary till we
+          reach the mailbox capacity.*/
+
+        String folderName = imapFolder.getName();
+        Date earliestDate = account.getEarliestPollDate();
+        final AtomicInteger headerProgress = new AtomicInteger(0);
+
+        int newLocalMessageCount = remoteMessagesToDownload.size() + localFolder.getMessageCount();
+        int oldMessagesStart = syncHelper.getRemoteStart(localFolder, imapFolder);
+        int oldMessagesEnd = oldMessagesStart - 1 + localFolder.getVisibleLimit() - newLocalMessageCount;
+
+        if (oldMessagesEnd > oldMessagesStart) {
+            List<ImapMessage> oldMessages = imapFolder.getMessages(oldMessagesStart, oldMessagesEnd, earliestDate, null);
```

Comment:
Looks like I made a mistake here.

The SELECT response for QRESYNC gives the UID + flags for any new messages that arrived since the last sync. However, in this line, we fetch only the UIDs (via UID SEARCH) of old messages that need to be need to be downloaded to fill the mailbox. 
This means that the flags for old messages are never downloaded.

I think the best way to fix this would be to use a FETCH command here instead of a UID SEARCH to fetch the UIDs and flags of old messages using their sequence numbers. This will mean adding support for FETCH (we only support UID FETCH currently).

Is there a better way ? @Valodim